src/expressions/capacity_margin.jl

```diff
@@ -0,0 +1,182 @@
+                    m;
+                    node=n,
+                    direction=direction(:to_node),
+                    stochastic_scenario=s,
+                    t=t,
+                    temporal_block=anything,
+                )
+                if is_storage_unit(u);
+                init=0,
+            )           
+
+            # Conventional and Renewable Capacity
+            + sum(
+                + unit_capacity[(unit=u, node=n_, direction=d, stochastic_scenario=s, analysis_time=t0, t=t)]
+                * unit_availability_factor[(unit=u, stochastic_scenario=s, analysis_time=t0, t=t)]
+                * (                   
+                    + sum(
+                        + units_available[u, s, t_ua]
+                        for (u, s, t_ua) in units_on_indices(
+                            m;
+                            unit=u,
+                            stochastic_scenario=s,
+                            t=t_overlaps_t(m; t=t),
+                            temporal_block=anything,
+                        );
+                        init=0,
+                    )                                           
+                )
+                for (u, n_, d) in indices(unit_capacity; node=n, direction=direction(:to_node)) if !is_storage_unit(u)
+            )
+        )
+        for (n, s, t) in expression_capacity_margin_indices(m)
+    )
+end
+
+"""
+    expression_capacity_margin_indices!(m::Model; t_range)
+
+    Return the indices for the capacity_margin expression
+"""
+
+function expression_capacity_margin_indices(m::Model; t_range=anything)
+    unique(
+        (node=n, stochastic_path=path, t=t)        
+        for n in indices(min_capacity_margin)        
+        for (n, t) in node_time_indices(m; node=n)
+        for path in active_stochastic_paths(
+            m,  
+            [                                      
+                node_stochastic_time_indices(m; node=n, t=t);
+                [
+                    (unit=u, stochastic_scenario=s, t=t2)
+                    for u in indices(unit_capacity; node=n, direction=direction(:to_node)) if !is_storage_unit(u)
+                    for (u, s, t2) in units_on_indices(m; unit=u, t=t_overlaps_t(m; t=t))
+                ];
+                [
+                   (unit=u, node=n, stochastic_scenario=s, t=t)
+                   for u in unit__to_node(node=n, direction=direction(:to_node)) if is_storage_unit(u)
+                   for (u, n, d, s, t) in unit_flow_indices(m; unit=u, node=n, direction=direction(:to_node), t=t) 
+                ];
+                [
+                   (unit=u, node=n, stochastic_scenario=s, t=t)
+                   for u in unit__to_node(node=n, direction=direction(:from_node)) if is_storage_unit(u)
+                   for (u, n, d, s, t) in unit_flow_indices(m; unit=u, node=n, direction=direction(:from_node), t=t) 
+                ]
+            ]
+            
+        )
+        
+    )
+end
```

Comment:
Probably a good idea to test performance in a large model...

Reply:
I'm just realising that I don't need to scan the unit_flow variables for active paths since they have to have the same temporal and stochastic structure as node_stochastic_time_indices. So this will be better.